File: 01.patterns/pattern-07.js
// this will output this
//         *
//       * * *
//     * * * * *
//   * * * * * * *
// * * * * * * * * *

// we can actually see the pattern here.. 
// [4, 1, 4]
// [3, 3, 3]
// [2, 5, 2]
// [1, 7, 1]
// [0, 9, 0]
// for conection with rows from column
// we can see that there is actually reverse of it, which is n - i - 1 for spaces
// we can see that for the stars it is actually 2 * i + 1

let n = 5
for(let i = 0; i < n; i++){
    let space1 = ""
    let stars = ""
    let space2 = ""

    for(let j = 0; j < n - i - 1; j++){
        space1 += "  "
        space2 += "  "
    }

    for(let k = 0; k < 2 * i + 1; k++){
        stars += "* "
    }

    console.log(space1 + stars + space2)
}

console.log("--- revision starts here ---")


//         *
//       * * *
//     * * * * *
//   * * * * * * *
// * * * * * * * * *
let k = 5
for(let i = 0; i < k; i++){
    // space1
    let space1 = ""
    for(let j = 0; j < k - i - 1; j++){
        space1 += "  "
    }

    let star = ""
    for(let n = 0; n < 2 * i + 1; n++){
        star += "* "
    }
    console.log(space1 + star)
}

// final revision
//         *
//       * * *
//     * * * * *
//   * * * * * * *
// * * * * * * * * *
let rows = 5
for(let i = 0; i < rows; i++){
    let space = ''
    for(let j = 0; j < rows - i - 1; j++){
        space += "  "
    }

    let stars = ''
    for(let j = 0; j < 2 * i + 1; j++){
        stars += '* '
    }

    console.log(space + stars + space)
}

File: 01.patterns/pattern-07.go
// ------------------ revision with js and go -------------------------
package main

import (
	"fmt"
	"strings"
)

//         *
//       * * *
//     * * * * *
//   * * * * * * *
// * * * * * * * * *

// 4 1 4
// 3 3 3
// 2 5 2
// 1 7 0
// 0 9 0

func main() {
	p := 5

	// for i := 0; i < p; i++ {

	// }

	for i := 0; i < p; i++ {
		space := strings.Repeat("  ", p-i-1)
		stars := strings.Repeat("* ", 2*i+1)
		fmt.Println(space + stars)
	}

	fmt.Println("---------")

	for i := 0; i < p; i++ {
		var b strings.Builder
		// b.Grow(3*p - 1)
		// space
		b.Grow((p-i-1)*2 + (2*i+1)*2)
		for j := 0; j < p-i-1; j++ {
			b.WriteString("  ")
		}

		for j := 0; j < 2*i+1; j++ {
			b.WriteString("* ")
		}

		fmt.Println(b.String())
	}
}
